### PR TITLE
Do not trim unsafe commits when open readonly engine

### PR DESCRIPTION
**Pull request by dnhatn**
Today we always trim unsafe commits (whose max_seq_no >= global checkpoint) before starting a read-write or read-only engine. This is mandatory for read-write engines because they must start with the safe commit. This is also fine for read-only engines since most of the cases we should have exactly one commit after closing an index (trimming is a noop). However, this is not safe for following indices which might have more than one commits when they are being closed.

With this change, we won't trim anything if we are going to open a read-only engine.

I will work on a follow up to skip resync/rollback if an index is closed and another follow-up to enable the sanity check in ReadOnly for following indices.

Relates #33888

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java [hunk 1]
```diff
@@ -103,22 +102,7 @@ public ReadOnlyEngine(EngineConfig config, SeqNoStats seqNoStats, TranslogStats
...
-                    final Version indexVersionCreated = engineConfig.getIndexSettings().getIndexVersionCreated();
-                    if (indexVersionCreated.onOrAfter(Version.V_8_0_0) ||
-                        (globalCheckpoint != SequenceNumbers.UNASSIGNED_SEQ_NO && indexVersionCreated.onOrAfter(Version.V_6_7_0))) {
-                        if (seqNoStats.getMaxSeqNo() != globalCheckpoint) {
-                            assertMaxSeqNoEqualsToGlobalCheckpoint(seqNoStats.getMaxSeqNo(), globalCheckpoint);
-                            throw new IllegalStateException("Maximum sequence number [" + seqNoStats.getMaxSeqNo()
-                                + "] from last commit does not match global checkpoint [" + globalCheckpoint + "]");
-                        }
-                    }
+                    ensureMaxSeqNoEqualsToGlobalCheckpoint(seqNoStats);
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/engine/ReadOnlyEngine.java]: I am working to make this check allow gaps in sequence numbers.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 2]
```diff
@@ -1439,12 +1439,15 @@ private void innerOpenEngineAndTranslog() throws IOException {
         assert recoveryState.getRecoverySource().expectEmptyRetentionLeases() == false || getRetentionLeases().leases().isEmpty()
             : "expected empty set of retention leases with recovery source [" + recoveryState.getRecoverySource()
             + "] but got " + getRetentionLeases();
-        trimUnsafeCommits();
+        // don't trim anything if we are going to open a read only engine.
+        if (engineFactory.isReadOnlyEngineFactory() == false) {
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I think we should add an assertion below that ensures that if it's returning true that it's a subclass of ReadOnlyEngine.

**Review by s1monw — COMMENTED** [review]
left one comment

**Review by henningandersen — COMMENTED** [review]
Left one comment, I would prefer to keep the `EngineFactory` as is if possible.

I am not enough up to speed on the why closing a follower index is problematic wrt. read-only indices trimming indexes, will reach out to you about this.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 3]
```diff
@@ -1439,12 +1439,15 @@ private void innerOpenEngineAndTranslog() throws IOException {
...
+            trimUnsafeCommits();
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: +1 to move trimUnsafeCommits to the InternalEngine.

**Comment by dnhatn** [comment]
@s1monw and @henningandersen I've moved trimUnsafeCommits to InternalEngine. It's nicer I think. Can you please have another look? Thank you!

**Review by henningandersen — APPROVED** [review]
LGTM.

Thanks @dnhatn

**Review by s1monw — APPROVED** [review]
looks much better! thanks

**Comment by dnhatn** [comment]
@s1monw @henningandersen thanks for reviewing.

**Pull request merged**